Add proposal of cross resource group nodes for Azure

feiskyer opened the pull request:
Add proposal of cross resource group nodes for Azure.

/sig azure

/assign @brendandburns @khenidak @colemickens
/cc @justaugustus @andyzhangx @kubernetes/sig-azure-misc

liggitt commented on contributors/design-proposals/azure/cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,124 @@
+  - The name of resource group, which is used for Azure nodes.
+  - `on-prem`, which is used for on-prem nodes.
+
+When initializing nodes, these two labels should be set for kubelet by provisioning tools, e.g.
```
are these looking to add new labels that kubelets set on their own Node objects? xref #911 for reasons why this is not desirable.

feiskyer replied:
Also yep, it's for kubelet add labels for itself.  We can add those labels to the whitelist after #911.

Note that azure cloud provider would validate resource group by calling Azure API. If the value is not consistent with Azure, then errors would be reported (`InstanceNotFound`).

liggitt commented on contributors/design-proposals/azure/cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,124 @@
+
+## Design
+
+Instance metadata is a general way to fetch node information for Azure, but it doesn't work if cloud-controller-manager is used (`kubelet --cloud-provider=external`). So it won't be used in this proposal. Instead, following two labels are proposed for providing required information:
```
use of the external cloud provider typically involves a controller that pulls information from metadata and sets it on Node objects. Using a source of truth outside the individual kubelets is preferable for setting topology information on Node objects, since it doesn't require trusting all nodes to categorize themselves appropriately.

liggitt commented on contributors/design-proposals/azure/cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,124 @@
+- `alpha.service-controller.kubernetes.io/exclude-balancer=true`, which is used to exclude the node from load balancer
+- `kubernetes.azure.com/resource-group=<rg-name>`, which provides external RG and is used to get node information. Candidate values are
+  - The name of resource group, which is used for Azure nodes.
+  - `on-prem`, which is used for on-prem nodes.
```
is this a reserved name (is an invalid resource group name)?

feiskyer replied:
yep, it's a valid resource group name. Actually, all label values are valid resource group name. So it's a reserved name here.

liggitt commented on contributors/design-proposals/azure/cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,124 @@
+
+- `alpha.service-controller.kubernetes.io/exclude-balancer=true`, which is used to exclude the node from load balancer
+- `kubernetes.azure.com/resource-group=<rg-name>`, which provides external RG and is used to get node information. Candidate values are
+  - The name of resource group, which is used for Azure nodes.
```
are all resource group names valid label values?

feiskyer replied:
Thanks for reminding this. 

No, resource group name is in length 1-90, while label value only accepts 63. I should add a limitation part for this.

justaugustus commented:
@feiskyer -- this should be submitted as a KEP. 
Can you move it `/keps/sig-azure` and make sure the proposal adheres to the [KEP template](https://github.com/kubernetes/community/blob/master/keps/0000-kep-template.md)?

Additionally, if we're planning work for this in the 1.12 Release cycle you'll need to:
- Submit a Feature issue detailing any work planned and link to this KEP
- File an [Exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) for that work to be accepted in the 1.12 cycle

/hold

feiskyer commented:
@justaugustus it's already tracked by feature https://github.com/kubernetes/features/issues/604. Let me convert this to KEP tomorrow.

justaugustus commented:
@feiskyer -- thanks for moving this over (and for already having the feature issue opened)! :)
I'll try and review the proposal tomorrow.

feiskyer commented:
/hold cancel

feiskyer commented:
ping @justaugustus  @khenidak Any new comments?

kkmsft commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+
+### Assumptions
+
+While news nodes (either from different RGs or on-prem) would be supported in this proposal, not all features would be supported for them. For example, AzureDisk will not work for on-prem nodes.
```
nit: news => new

feiskyer replied:
Fixed.

kkmsft commented:
/lgtm

khenidak commented:
/lgtm

justaugustus commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+
+Today, Azure cloud provider only supports nodes from specified RG (which is set in cloud provider configure file). For nodes in different RG, Azure cloud provider reports `InstanceNotFound` error and thus they would be removed by controller manager. So as on-prem nodes.
+
+With managed cluster (limited access to nodes) users may need nodes where they can customize in ways that are not possible in a managed service. So this document proposes to support joining arbitrary nodes to the cluster and make required changes for both Azure cloud provider and provision setups, which includes
```
With managed clusters, like [AKS](https://docs.microsoft.com/en-us/azure/aks/), there is limited access to configure nodes. There are instances where users may need to customize nodes in ways that are not possible in a managed service.**

justaugustus replied:
This document proposes support for joining arbitrary nodes to a cluster and the required changes to make in both the Azure cloud provider and provisioned setups, which include:**

justaugustus commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+
+With managed cluster (limited access to nodes) users may need nodes where they can customize in ways that are not possible in a managed service. So this document proposes to support joining arbitrary nodes to the cluster and make required changes for both Azure cloud provider and provision setups, which includes
+
+- Provision tools should setup kubelet with required labels (e.g. via `--node-labels`)
```
Provisioning**

justaugustus commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+
+- Nodes are in same region and set with required labels (as clarified in the following design part)
+- Nodes will not be part of the load balancer managed by cloud provider
+- Both node and container networking have setup properly
```
s/have setup properly/are properly configured

justaugustus commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+- Both node and container networking have setup properly
+- AzureDisk is supported for Azure cross-RG nodes, but not for on-prem nodes
+
+Besides, feature gate [ServiceNodeExclusion](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L174) should also be enabled for Kubernetes cluster.
```
s/Besides,/In addition,

justaugustus replied:
s/should/must

justaugustus commented on keps/sig-azure/0024-20180809-cross-resource-group-nodes.md:
```diff
@@ -0,0 +1,165 @@
+
+### Non-Goals
+
+Note that provisioning Kubernetes cluster, setting up networking and provisioning new nodes are out of this proposal scope. Those could be done by external provisioning tools (e.g. acs-engine).
```
provisioning the**

feiskyer commented:
Offline chat with @justaugustus, we have agreed to use label `kubernetes.azure.com/on-prem=true` for on-prem nodes. In this way, we don't need to reserve some names for resource group names.

@justaugustus Addressed comments, PTAL

justaugustus commented:
Perfect. Thanks for addressing all of my nits, @feiskyer!
/lgtm
/approve

justaugustus commented:
/assign @calebamiles @idvoretskyi @jdumars 
(...although, shouldn't I be able to merge myself as there's an [OWNERS](https://github.com/kubernetes/community/blob/master/keps/sig-azure/OWNERS) file in `/keps/sig-azure`?)

feiskyer commented:
/hold cancel

idvoretskyi commented:
@justaugustus you should be able to.

/uncc

justaugustus commented:
/approve
(crosses fingers)

feiskyer commented:
@idvoretskyi @justaugustus Thank you very much.
